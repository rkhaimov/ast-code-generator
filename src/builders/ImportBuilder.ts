import { ImportDeclarationStructure } from 'ts-simple-ast';

import { ProjectManager } from '../managers/ProjectManager';

class _ImportBuilder {
  importRepositoryBase(): ImportDeclarationStructure {
    return {
      moduleSpecifier: `./${ProjectManager.FILE_NAMES.REPOSITORY_BASE}`,
      namedImports: [{ name: ProjectManager.FILE_NAMES.REPOSITORY_BASE }],
    };
  }
}

export const ImportBuilder = new _ImportBuilder();
