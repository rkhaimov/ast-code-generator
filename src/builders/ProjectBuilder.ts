export class ProjectBuilder {
  static FILE_NAMES = {
    DEFINITIONS: 'definitions',
    REPOSITORIES: 'repositories',
    REPOSITORY_BASE: 'RepositoryBase',
  };
}
