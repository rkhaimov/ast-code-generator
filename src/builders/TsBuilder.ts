import { ts } from 'ts-simple-ast';

class _TsBuilder {
  project = ts.createSourceFile('ts-build-source.ts', '', ts.ScriptTarget.ES2015, true, ts.ScriptKind.TS);
  printer = ts.createPrinter();

  print(node: ts.Node): string {
    return this.printer.printNode(ts.EmitHint.Unspecified, node, this.project);
  }
}

export const TsBuilder = new _TsBuilder();
