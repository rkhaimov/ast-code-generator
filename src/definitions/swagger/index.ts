interface ISwagger {
  swagger: string;
  info: {
    version: string;
    title: string;
  };
  paths: {
    [path: string]: Partial<ISwaggerPath>;
  };
  definitions: {
    [model: string]: ISwaggerDefinition;
  };
  securityDefinitions: {};
  security: object[];
}

interface ISwaggerPath {
  get: ISwaggerOperation;
  post: ISwaggerOperation;
}

interface ISwaggerOperation {
  tags: string[];
  operationId: string;
  consumes: string[];
  produces: string[];
  parameters: Array<{
    name: string;
    in: string;
    required: boolean;
    schema: ISwaggerDefinitionRef | ISwaggerDefinition;
  }>;
  responses: {
    200: {
      description: string;
      schema: ISwaggerDefinitionRef | ISwaggerDefinition;
    };
  };
}

interface ISwaggerDefinitionRef {
  $ref: string;
}

interface ISwaggerDefinition {
  type: 'object';
  properties: {
    [property: string]: SwaggerDefinitionPropertyTypes;
  };
}

type SwaggerDefinitionPropertyTypes = ISwaggerString | ISwaggerInt | ISwaggerBoolean | ISwaggerArray;

interface ISwaggerPropertyTypeBase {
  readOnly?: boolean;
}

interface ISwaggerString extends ISwaggerPropertyTypeBase {
  type: 'string';
  enum?: string[];
  format?: 'date-time';
}

interface ISwaggerInt extends ISwaggerPropertyTypeBase {
  type: 'integer';
  format: 'int32';
}

interface ISwaggerBoolean extends ISwaggerPropertyTypeBase {
  type: 'boolean';
}

interface ISwaggerArray extends ISwaggerPropertyTypeBase  {
  type: 'array';
  uniqueItems: boolean;
  items: ISwaggerDefinitionRef | ISwaggerDefinition;
}
