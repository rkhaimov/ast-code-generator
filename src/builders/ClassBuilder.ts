import reduce from 'lodash/reduce';
import last from 'lodash/last';
import isEmpty from 'lodash/isEmpty';

import { ISwagger, ISwaggerMethod, ISwaggerMethodParam, ISwaggerMethods } from '../definitions/swagger';
import {
  FunctionExpressionBodyTypes,
  IClassDeclaration,
  IFunctionParam,
  IMethodDefinition,
} from '../definitions/class/ast';

interface IClassBuilder {
  build(name: string, methods: ISwagger['paths']): IClassDeclaration;
}

class _ClassBuilder implements IClassBuilder {
  build(name: string, methods: ISwagger['paths']): IClassDeclaration {
    const body = reduce(methods, (acc, operation, api) => {
      const method = this.buildMethod(api, operation);

      return acc.concat(method);
    }, [] as IMethodDefinition[]);

    return {
      type: 'ClassDeclaration',
      id: {
        type: 'Identifier',
        name: `${name}Repository`,
      },
      superClass: {
        type: 'Identifier',
        name: 'RepositoryBase', // TODO: Replace to class name
      },
      body: {
        type: 'ClassBody',
        body,
      },
    };
  }

  private buildMethod(api: string, operations: ISwaggerMethods): IMethodDefinition {
    const method: ISwaggerMethod = 'get' in operations ? operations.get : operations.post;
    const methodName = last<string>(api.split('/'));

    return {
      type: 'MethodDefinition',
      kind: 'method',
      static: false,
      computed: false,
      key: {
        name: methodName!,
        type: 'Identifier',
      },
      value: {
        type: 'FunctionExpression',
        params: this.buildFunctionParams(method.parameters),
        body: {
          type: 'BlockStatement',
          body: this.buildFunctionImplementation(api, operations),
        },
        async: false,
        generator: false,
        expression: false,
        id: null,
      },
    };
  }

  private buildFunctionParams(swaggerParams: ISwaggerMethodParam[]): IFunctionParam[] {
    if (isEmpty(swaggerParams)) {
      return [];
    }

    const param = {
      name: 'payload',
      type: 'Identifier',
    };

    return [param as IFunctionParam];
  }

  private buildFunctionImplementation(api: string, operations: ISwaggerMethods): FunctionExpressionBodyTypes[] {
    return [];
  }
}

export const ClassBuilder: IClassBuilder = new _ClassBuilder();
