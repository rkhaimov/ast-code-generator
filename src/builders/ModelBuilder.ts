/* tslint:disable:max-line-length */
import { reduce, last } from 'lodash';
import { InterfaceDeclarationStructure, PropertySignatureStructure } from 'ts-simple-ast';

import { ISwagger, SwaggerDefinitionPropertyTypes } from '../definitions/swagger';

class _ModelBuilder {
  buildModel(model: string, definition: ISwagger['definitions'][string]): InterfaceDeclarationStructure {
    const properties: InterfaceDeclarationStructure['properties'] = reduce(definition.properties, (acc, value: SwaggerDefinitionPropertyTypes, name) => {
      const property: PropertySignatureStructure = {
        name,
        type: this.toTsType(value),
      };

      return (acc as InterfaceDeclarationStructure['properties'])!.concat(property);
    }, [] as InterfaceDeclarationStructure['properties']);

    return {
      name: `I${model}`,
      properties,
      isExported: true,
    };
  }

  private toTsType(property: SwaggerDefinitionPropertyTypes): string {
    switch (property.type) {
      case 'integer': {
        return 'number';
      }
      case 'string': {
        if (property.enum) {
          return property.enum.map((value) => `'${value}'`).join(' | ');
        }

        return 'string';
      }
      case 'array': {
        if (!('$ref' in property.items)) {
          throw new Error('Unrecognizable array swagger structure');
        }

        const name = this.getRefDefinitionName(property.items.$ref);

        return `${name}[]`;
      }
      default: {
        return property.type;
      }
    }
  }

  private getRefDefinitionName(ref: string): string {
    return `I${last(ref.split('/'))}`;
  }
}

export const ModelBuilder = new _ModelBuilder();
