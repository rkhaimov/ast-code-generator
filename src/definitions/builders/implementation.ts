import { ISwaggerMethodParam } from '../swagger';

export type ArgumentsGroup = {
  [P in ISwaggerMethodParam['in']]: ISwaggerMethodParam;
};
