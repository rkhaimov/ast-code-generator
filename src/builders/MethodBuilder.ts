import { isEmpty, last, keys, first, get, lowerFirst, isNil } from 'lodash';
import { MethodDeclarationStructure, ParameterDeclarationStructure, ts } from 'ts-simple-ast';
import escodegen from 'escodegen';

import { GetImplementationBuilder } from './implementation/methods/GetImplementationBuilder';
import { PostImplementationBuilder } from './implementation/methods/PostImplementationBuilder';
import { PutImplementationBuilder } from './implementation/methods/PutImplementationBuilder';

import { BlockStatementBody } from '../definitions/ast/common';
import {
  ISwaggerMethod,
  SwaggerMethodParam,
  ISwaggerOperations,
  SwaggerDefinitionPropertyTypes,
} from '../definitions/swagger';
import { ModelBuilder } from './ModelBuilder';
import { TsBuilder } from './TsBuilder';

interface IMethodBuilder {
  buildMethod(api: string, operations: ISwaggerOperations): MethodDeclarationStructure;
}

class _MethodBuilder implements IMethodBuilder {
  buildMethod(api: string, operations: ISwaggerOperations): MethodDeclarationStructure {
    const methodName = last<string>(api.split('/'));
    const methods: string[] = keys(operations);
    const method: ISwaggerMethod = get(operations, first(methods)!);

    return {
      name: methodName!,
      parameters: this.buildFunctionParams(method.parameters),
      bodyText: this.buildBodyText(api, operations),
      returnType: this.getReturnType(method.responses),
    };
  }

  private getReturnType(response: ISwaggerMethod['responses']): string {
    if (isNil(response['200'].schema)) {
      return TsBuilder.print(ts.createTypeReferenceNode('Promise', [ts.createTypeReferenceNode('void', [])]));
    }

    if ('$ref' in response['200'].schema) {
      const name = ModelBuilder.getRefDefinitionName(response['200'].schema.$ref);

      return TsBuilder.print(ts.createTypeReferenceNode('Promise', [ts.createTypeReferenceNode(name, [])]));
    }

    const type = ModelBuilder.toTsType(response['200'].schema);

    return TsBuilder.print(ts.createTypeReferenceNode('Promise', [ts.createTypeReferenceNode(type, [])]));
  }

  private buildBodyText(api: string, operations: ISwaggerOperations) {
    return escodegen.generate(this.buildFunctionBody(api, operations));
  }

  private buildFunctionBody(api: string, operations: ISwaggerOperations): BlockStatementBody {
    if ('get' in operations) {
      return GetImplementationBuilder.buildImplementation(api, operations.get);
    }

    if ('post' in operations) {
      return PostImplementationBuilder.buildImplementation(api, operations.post);
    }

    if ('put' in operations) {
      return PutImplementationBuilder.buildImplementation(api, operations.put);
    }

    if ('delete' in operations) {
      return PutImplementationBuilder.buildImplementation(api, operations.delete);
    }

    throw new Error(`Unknown operation(s) was provided ${keys(operations)}`);
  }

  private buildFunctionParams(swaggerParams: SwaggerMethodParam[]): ParameterDeclarationStructure[] {
    if (isEmpty(swaggerParams)) {
      return [];
    }

    const literalTypeNodes = swaggerParams.map<ts.TypeElement>((parameter) => {
      const typeName = ModelBuilder.toTsType(this.convertToSwaggerDefinition(parameter));
      const questionToken = parameter.required ? undefined : ts.createToken(ts.SyntaxKind.QuestionToken);

      return ts.createPropertySignature(
        [],
        lowerFirst(parameter.name),
        questionToken,
        ts.createTypeReferenceNode(typeName, []),
        undefined,
      );
    });

    const param: ParameterDeclarationStructure = {
      name: 'payload',
      type: TsBuilder.print(ts.createTypeLiteralNode(literalTypeNodes)),
    };

    return [param];
  }

  private convertToSwaggerDefinition(param: SwaggerMethodParam): SwaggerDefinitionPropertyTypes {
    switch (param.in) {
      case 'body': {
        return {
          type: 'object',
          properties: param.schema,
        };
      }
      case 'path': {
        return {
          ...param as SwaggerDefinitionPropertyTypes,
          type: 'string',
          format: param.format === 'int32' ? undefined : param.format,
        };
      }
      case 'query': {
        return param as SwaggerDefinitionPropertyTypes;
      }
      default: {
        return this.throwIfNever(param, '');
      }
    }
  }

  private throwIfNever(arg: never, message: string): never {
    throw new Error(message);
  }
}

export const MethodBuilder: IMethodBuilder = new _MethodBuilder();
