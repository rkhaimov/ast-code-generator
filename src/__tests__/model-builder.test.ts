import { Project, SourceFile } from 'ts-simple-ast';

import { ModelBuilder } from '../builders/ModelBuilder';

import swaggerMock from './__mocks__/swagger.json';

import { ISwagger } from '../definitions/swagger';

describe('Model builder works well when', () => {
  const project = new Project();
  const models: SourceFile = project.createSourceFile('models');

  it('matches giving code style', () => {
    const model = ModelBuilder.buildModel(
      'AccountInfo',
      swaggerMock.definitions.AccountInfo as ISwagger['definitions'][string],
    );

    models.addInterface(model);

    expect(models.getFullText()).toMatchSnapshot();
  });
});
