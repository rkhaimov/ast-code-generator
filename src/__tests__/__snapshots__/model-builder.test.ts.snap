// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model builder works well when matches giving code style 1`] = `
"export interface IAccountInfo {
    userName: string;
    firstName: string;
    lastName: string;
}

export interface ISignInRequestModel {
    user: string;
    password: string;
}

export interface ISignInResponseModel {
    success: boolean;
    resultCode: 'Success' | 'InvalidLoginOrPassword';
    accessToken: string;
    refreshToken: string;
}

export interface IAgreementModel {
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    initiator: string;
    deadline: string;
    status: string;
    id: number;
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
    assignedTo: string;
}

export interface ITaskCountModel {
    kind: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    count: number;
}

export interface IDetailsAgreementModel {
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    company: string;
    signer: string;
    signerEmail: string;
    note: string;
    initiator: string;
    iniciatorEmail: string;
    departmentBoss: string;
    departmentBossEmail: string;
    deadLine: string;
    afterAgreement: string;
    isSignAndSend: boolean;
    sendAddress: string;
    status: string;
    commentary: string;
    leadURL: string;
    agreementNo: string;
    docId: string;
    agreementType: string;
    department: string;
    receiveDate: string;
    project: string;
    contractDate: string;
    counterparty: string;
    currency: string;
    isRegisrty: boolean;
    isSendToAgreement: boolean;
    amount: string;
    paymentPlan: boolean;
    paymentStartDate: string;
    paymentEndDate: string;
    contractType: string;
    mainAgreement: string;
    attachments: IAttachment[];
    agreementTitle: string;
    agreementStatus: string;
    assignedToEmail: string;
    agreementId: number;
    editorEmail: string;
    id: number;
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
    assignedTo: string;
}

export interface IAttachment {
    fileName: string;
    linkUrl: string;
}

export interface ITaskOutcome {
    name: string;
    description: string;
}

export interface ITaskParticipant {
    title: string;
    userName: string;
    position: string;
    email: string;
}

export interface IDelegateTaskModel {
    userName: string;
    comments: string;
}

export interface IProcessingTaskModel {
    outcomeName: string;
    comments: string;
}

export interface IOutsourcingJob {
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    workflowName: string;
    id: number;
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
    status: string;
    assignedTo: string;
}

export interface IProcessingJobTaskModel {
    hours: number;
    outcomeName: string;
    comments: string;
}

export interface IOvertimeJob {
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    workflowName: string;
    id: number;
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
    status: string;
    assignedTo: string;
}

export interface IPaymentRequest {
    payUntilDate: string;
    budgetCode: string;
    currencyCode: string;
    supplier: string;
    paymentMethod: string;
    invNo: string;
    amount: number;
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    id: number;
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
    status: string;
    assignedTo: string;
}

export interface ITaskModel {
    id: number;
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
}

export interface IAppUser {
    userName: string;
    surname: string;
    password: string;
}

export interface IUserListItem {
    firstName: string;
    lastName: string;
    email: string;
}

export interface IVacation {
    taskType: 'AllTasks' | 'Unknown' | 'Vacation' | 'Outsourcing' | 'Overtime' | 'Agreement' | 'PaymentRequest';
    id: number;
    title: string;
    created: string;
    creatorEmail: string;
    creatorFirstName: string;
    creatorLastName: string;
    modified: string;
    editor: string;
    status: string;
    assignedTo: string;
}
"
`;
